get-qualify-export-donor-data/qualify-data/qualify-data.py

```diff
@@ -134,21 +157,72 @@ def removeInvalidCgmValues(df):
 
 def removeDuplicates(df, criteriaDF):
     nBefore = len(df)
-    df = df.loc[~(criteriaDF.duplicated())]
+    df = df.loc[~(df[criteriaDF].duplicated())]
     df = df.reset_index(drop=True)
     nDuplicatesRemoved = nBefore - len(df)
 
     return df, nDuplicatesRemoved
 
 
-def roundTimeAndDropDups(df, timeInterval):
-    df["roundedTime"] = \
-        pd.DatetimeIndex(df["time"]).round(str(timeInterval) + "min")
+def removeCgmDuplicates(df, timeCriterion):
+    if timeCriterion in df:
+        df.sort_values(by=[timeCriterion, "uploadTime"],
+                       ascending=[False, False],
+                       inplace=True)
+
+        dfIsNull = df[df[timeCriterion].isnull()]
+        dfNotNull = df[df[timeCriterion].notnull()]
+
+        dfNotNull, nDuplicatesRemoved = \
+            removeDuplicates(dfNotNull, [timeCriterion, "value"])
+
+        df = pd.concat([dfIsNull, dfNotNull])
+
+        df.sort_values(by=[timeCriterion, "uploadTime"],
+                   ascending=[False, False],
+                   inplace=True)
+    else:
+        nDuplicatesRemoved = 0
 
-    df, nDuplicatesRemoved = removeDuplicates(df, df["roundedTime"])
     return df, nDuplicatesRemoved
 
 
+def roundCgm5Minutes(df):
```

Comment:
Why do you need to do this? The drivers make sure that the basal profile is contiguous. Would this cause a possibility of non-contiguous data?
Also, it's not factual - the data that comes from the 670G is when the micro-boluses occur, so changing the times is effectively altering the data.

Reply:
good question @pazaan, the rounding only applies to cgm data, and was implemented because of some UTCbootstrapping errors that affect the data. Also, for now, the rounding of the times only applies to the qualification scripts to determine if the data meets big data partners qualification standards. When we export the data, we do not round the data.

Reply:
Herp de derp. The clue was in the `CGM` name 🤦‍♂️. But thanks for clarifying that it only applies to the qualification, and not the actual data sets.